### DISCUSSION TITLE
INDENT/DEDENT for YAML Lists?

### DISCUSSION THREAD

**Discussion by blewis13** [eclipse-langium/langium#1696] [Q&A] [answered]
My whitespace language grammar patterns I've used in Xtext for years for YAML similar grammars does not appear to jive with Langium's 3.2.0 implementation of indendation.

I'm trying to get YAML lists as:

```
my_list:
  - item1
  - item2
my_next_list:
  - item3
  - item4
```

.. and it's not working _at all_.  I've confirmed npm has langium 3.2.0 installed. yo generated the extension folder with 3.2.0.  My VSCode langium extension is at 3.1.1 (newer not available but grammar highlights fine without errors). Grammar generates and builds fine. Launch debug VSCode and it does not work as expected.

The grammar:

```
RepoSpec:
    'name' ':' name=ID
    'boms' ':' 
        INDENT
        ('-' bomPaths+=PathString)+
        DEDENT
    ('tools' ':'
        INDENT
        ('-' toolPaths+=PathString)+
        DEDENT)?
    ;

PathString returns string:
  (STRING | ID_or_KW | '$' | '/' | '.' | '..' | INT)*;

hidden terminal NL: /[\r\n]+/;
terminal INDENT: 'synthetic:indent'; // The content doesn't really matter.
terminal DEDENT: 'synthetic:dedent'; // It will be overridden anyway
hidden terminal WS: /[\t ]+/; // whitespace and newlines now have to be separated

terminal ID: /[_a-zA-Z][\w_]*/;
terminal INT returns number: /[0-9]+/;
terminal STRING: /"(\\.|[^"\\])*"|'(\\.|[^'\\])*'/;

hidden terminal SL_COMMENT: /#[^\n\r]*/;
```

And basically every single whitespace item possible is failing when I run this in debug mode:

With this text:

```
name: foo
boms: 
  - src/foo/config/bom.yml
tools: 
  - src/foo/config/tools


```

I get the '-' after boms highlighted as error with: Expecting token of type 'INDENT' but found `-`.
The ':' after tools errors with:  Expecting token of type 'DEDENT' but found `:`.
Right after the last line config/tools which is follow with a newline, I get the error: Expecting token of type 'DEDENT' but found ``.

**Comment by aabounegm** [comment, marked as answer]
Hi @blewis13, did you remember to register the `IndentationAwareTokenBuilder` and `Lexer` in your parser services?
The instructions to do so are available in this (yet to be published) recipe: https://github.com/eclipse-langium/langium-website/pull/246

**Comment by blewis13** [comment]
Yes, that was what I was missing and the link to the recipe explained it well. Thanks for your help!
